### PR TITLE
os/dbuild.sh : Add usbrule menu in download option

### PR DESCRIPTION
**Pull request by hasw7569**
I added usbrule menu for user convenient
If user selectes usbrule menu, then it will automatically creates a target board-specific .rule file in /etc/udev/rule.d directory.

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [db707410bafd16e4cd86dffcf499a47f91c57631] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [db707410bafd16e4cd86dffcf499a47f91c57631] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [398ec11f24d0eea755eef879a58361c419bb3ed4] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [398ec11f24d0eea755eef879a58361c419bb3ed4] - Code Rule Check (C++) OK.

File: os/dbuild.sh [hunk 1]
```diff
@@ -222,6 +228,7 @@ function SELECT_BOARD()
 	fi
 
 	if [ ! -z ${BOARD} ]; then
+		MAKE_RULES
```
Comment (sunghan-chang) [inline os/dbuild.sh]: I don't think this is necessary at this time. Let's move this at download.

File: os/dbuild.sh [hunk 2]
```diff
@@ -29,6 +29,12 @@ DOCKER_VERSION="1.5.1"
 STATUS_LIST="NOT_CONFIGURED BOARD_CONFIGURED CONFIGURED BUILT PREPARE_DL DOWNLOAD"
 BUILD_CMD=make
 
+rules_board=("artik" "cy4390x" "esp_wrover_kit" "imxrt" "stm32")
+rules_idVendor=("0403" "0a5c" "0403" "0d28" "0483")
+rules_idProduct=("6010" "43fa" "6010" "0204" "3748")
+rules_STLink_idVendor="0483"
+rules_STLink_idProduct=("3752" "3753" "374d" "374e" "374f")
```
Comment (sunghan-chang) [inline os/dbuild.sh]: Let's discuss how to manage the values.

**Comment by seinfra** [comment]
Target : [22c3d467cdba4482c433d0c4d2e432a3221d620e] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [22c3d467cdba4482c433d0c4d2e432a3221d620e] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [8529dcbccfe44f3b25cb931cd028d8aac19e68ee] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [8529dcbccfe44f3b25cb931cd028d8aac19e68ee] - Code Rule Check (C++) OK.

File: build/configs/artik053/.flashSpec.xml [hunk 3]
```diff
@@ -22,6 +22,7 @@
     <options>
         <option desc="Flash all binaries and images">ALL</option>
         <option desc="Flash TizenRT binary">OS</option>
+	<option desc="Automatic generate of usb rule">USBRULE</option>
```
Comment (sunghan-chang) [inline build/configs/artik053/.flashSpec.xml]: 1. alignment
2. "Register the USB rule"

File: build/tools/esp32/esptool_py/Makefile [hunk 4]
```diff
@@ -88,6 +89,9 @@ ALL: flash romfs
 
 OS: app-flash
 
+USBRULE: 
+		$(USBRULE_DIR)/esp32_DevKitC_usbrule.sh
```
Comment (sunghan-chang) [inline build/tools/esp32/esptool_py/Makefile]: 1. Please remove a tab.
2. esp32_usbrule.sh

File: build/configs/stm32l4r9ai-disco/.flashSpec.xml [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+ either express or implied. See the License for the specific
+ language governing permissions and limitations under the License.
+-->
+<flash name="stm32l4r9ai-disco flash" virtual="false">
+    <!-- path attribute value must be relative path from project root -->
+    <executors>
+        <execute name="flash" path="${PROJECT_PATH}/os" command="make download" os="linux" bit="all"></execute>
+    </executors>
+    <options>
+        <option desc="Flash all binaries and images">ALL</option>
```
Comment (sunghan-chang) [inline build/configs/stm32l4r9ai-disco/.flashSpec.xml]: Please remain "ALL" and "USBRULE".

File: build/configs/stm32f4xx/stm32f4xx_usbrule.sh [hunk 6]
```diff
@@ -0,0 +1,64 @@
+#!/bin/bash
```
Comment (sunghan-chang) [inline build/configs/stm32f4xx/stm32f4xx_usbrule.sh]: #!/usr/bin/env bash

File: build/configs/stm32f4xx/stm32f4xx_usbrule.sh [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        done
+        if [ $flag == $true ]; then
+            echo ".rules file already exist"
+            exit
+        fi
+    done
+    MAKE_RULE
+}
+
+function MAKE_RULE()
```
Comment (sunghan-chang) [inline build/configs/stm32f4xx/stm32f4xx_usbrule.sh]: REGISTER_RULE

File: build/configs/artik05x/artik05x_download.sh [hunk 8]
```diff
@@ -273,7 +281,7 @@ while test $# -gt 0; do
 		--verify)
 			VERIFY=verify
 			;;
-		ALL|OS|APPS|ROM|BL1|BL2|SSSFW|WLANFW|OTA|MICOM|WIFI|USER|all|os|apps|rom|bl1|bl2|sssfw|wlanfw|ota|micom|wifi|user)
+		ALL|OS|USBRULE|APPS|ROM|BL1|BL2|SSSFW|WLANFW|OTA|MICOM|WIFI|USER|all|os|apps|rom|bl1|bl2|sssfw|wlanfw|ota|micom|wifi|user)
```
Comment (sunghan-chang) [inline build/configs/artik05x/artik05x_download.sh]: Please split the USBRULE from other paritioin

File: build/configs/artik053/.flashSpec.xml (hunk 3 above)
Comment (hasw7569) [reply build/configs/artik053/.flashSpec.xml]: modified

File: build/tools/esp32/esptool_py/Makefile (hunk 4 above)
Comment (hasw7569) [reply build/tools/esp32/esptool_py/Makefile]: modified

File: build/configs/stm32l4r9ai-disco/.flashSpec.xml (hunk 5 above)
Comment (hasw7569) [reply build/configs/stm32l4r9ai-disco/.flashSpec.xml]: modified

File: build/configs/stm32f4xx/stm32f4xx_usbrule.sh (hunk 6 above)
Comment (hasw7569) [reply build/configs/stm32f4xx/stm32f4xx_usbrule.sh]: modified

File: build/configs/stm32f4xx/stm32f4xx_usbrule.sh (hunk 7 above)
Comment (hasw7569) [reply build/configs/stm32f4xx/stm32f4xx_usbrule.sh]: modified

File: build/configs/artik05x/artik05x_download.sh (hunk 8 above)
Comment (hasw7569) [reply build/configs/artik05x/artik05x_download.sh]: modified

**Comment by seinfra** [comment]
Target : [c1cec0ca158e126dbf713ed951041d76126cc7a0] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [c1cec0ca158e126dbf713ed951041d76126cc7a0] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [5723664ed3988acc8e3280cf4a06e320152cf89a] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [5723664ed3988acc8e3280cf4a06e320152cf89a] - Code Rule Check OK.

File: build/configs/artik05x/artik05x_usbrule.sh [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+###########################################################################
+#
+# File   : artik05x_usbrule.sh
+# Description : usb rule script for ARTIK 05X
+
+board="artik"
+idVendor="0403"
+idProduct="6010"
+
+function RULE_CHECK()
```
Comment (sunghan-chang) [inline build/configs/artik05x/artik05x_usbrule.sh]: CHECK_RULE

File: build/configs/artik05x/artik05x_usbrule.sh [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+# language governing permissions and limitations under the License.
+#
+###########################################################################
+#
+# File   : artik05x_usbrule.sh
+# Description : usb rule script for ARTIK 05X
+
+board="artik"
+idVendor="0403"
+idProduct="6010"
```
Comment (sunghan-chang) [inline build/configs/artik05x/artik05x_usbrule.sh]: Except those values, there is no difference between each board. How about making a common file and calling it with arguments?

**Comment by seinfra** [comment]
Target : [8ff9763485f6493b8ba2dcfff2d44fbb4eefecb3] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [8ff9763485f6493b8ba2dcfff2d44fbb4eefecb3] - Code Rule Check (C++) OK.

File: build/configs/artik05x/artik05x_usbrule.sh (hunk 9 above)
Comment (hasw7569) [reply build/configs/artik05x/artik05x_usbrule.sh]: modified

File: build/configs/artik05x/artik05x_usbrule.sh (hunk 10 above)
Comment (hasw7569) [reply build/configs/artik05x/artik05x_usbrule.sh]: modified

**Comment by seinfra** [comment]
Target : [ade2b6caf8ac3658acb7555cf435dc744ddb837a] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [ade2b6caf8ac3658acb7555cf435dc744ddb837a] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [890116bcb2f7dbebd9c96533a467a30444657bee] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [890116bcb2f7dbebd9c96533a467a30444657bee] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [1b11d9a6deb3d82a9ab72f4a4e6404d26c184427] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [1b11d9a6deb3d82a9ab72f4a4e6404d26c184427] - Code Rule Check OK.

**Pull request merged**